ipa-core/src/helpers/buffers/circular.rs

```diff
@@ -0,0 +1,788 @@
+use std::ops::RangeInclusive;
+
+use generic_array::GenericArray;
+use typenum::Unsigned;
+
+use crate::ff::Serializable;
+
+/// This is a specialized version of circular buffer implementation,
+/// tailored to what [`OrderingSender`] needs.
+///
+/// This construction requires one extra parameter, compared to
+/// traditional ring buffers: `read_size` that specifies the smallest
+/// continuous block of bytes that can be read off this buffer. It
+/// also requires the total `capacity` and the size of one write
+/// `write_size` to be provided at construction time. This allows it
+/// to use a single allocation.
+///
+/// For example: if buffer capacity is set to 32k and `read_size` is
+/// 4k, then buffer can only be read in blocks of 4k. `write_size` can
+/// be any value, aligned with 4k `read_size`.
+///
+/// The capacity of this buffer does not need to be a power of two, although
+/// you may want to have exactly that for performance of modulo operations
+/// widely used internally.
+///
+/// This allows using this buffer in scenarios where the working window
+/// is wide. If readers want smaller chunks of data, but can operate on
+/// them fast enough, then writers experience less interruptions hitting
+/// the capacity limit, because it can be set large enough.
+///
+/// This buffer can also be closed, using [`close`] method. After it is
+/// closed, it allows reads of any size, ubt it guarantees that all of them
+/// will be aligned with `write_size`.
+///
+/// ## Implementation notes
+/// This buffer is built over a [`Vec`] with two extra pointers that indicate
+/// the place to read from, and to write next chunk to. When read happens,
+/// it moves the read pointer until it meets the write pointer.
+/// When read points to the same location as write, this buffer is considered
+/// empty.
+///
+/// Both pointers operate within the range `[0, 2*capacity)` and clamped into
+/// the working range when used as index into the internal buffer. The reason
+/// is to be able to distinguish empty and full buffers.
+///
+/// This implementation does not perform checks in optimized builds,
+/// relying on [`OrderingSender`] to enforce correctness. If taken away,
+/// necessary adjustments need to be made to avoid data corruption.
+///
+/// ## Alternative implementations
+/// If alignment to `read_size` is too much, a [`BipBuffer`] can be used instead.
+///
+/// ## Future improvements
+/// [`OrderingSender`] currently synchronizes reader and writers, but it does not
+/// have to if this implementation is made thread-safe. There exists a well-known
+/// lock-free FIFO implementation for a single producer, single consumer that uses
+/// atomics for read and write pointers. We can't make use of it as is because there
+/// are more than one writer. However, [`OrderingSender`] already knows now to allow
+/// only one write at a time, so it could be possible to make the entire
```

Comment:
More than one write at a time?

Reply:
yea the whole sentence should be:  However, [`OrderingSender`] already knows **how to allow
 only one write** at a time